### PR TITLE
separate handler and middleware

### PR DESCRIPTION
**Pull request by instabledesign**

### CODE REVIEW

File: middleware/caller.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package middleware
+
+import (
+	"github.com/gol4ng/logger"
+	"runtime"
+)
+
+// This handler will add 2 fields in the handled entry context
```
Comment (rmasclef) [inline middleware/caller.go]: middleware

File: middleware/caller.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package middleware
+
+import (
+	"github.com/gol4ng/logger"
+	"runtime"
```
Comment (rmasclef) [inline middleware/caller.go]: sort imports (add blank line btw std lib and gol4ng pkg)

File: middleware/caller.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+)
+
+// This handler will add 2 fields in the handled entry context
+// eg:
+//     _file:/..../file_of_the_handle_caller.go
+//     _line:72
+func Caller(skip int) logger.Middleware {
```
Comment (rmasclef) [inline middleware/caller.go]: no interface returned here ?

File: middleware/caller.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	return func(handler logger.HandlerInterface) logger.HandlerInterface {
```
Comment (rmasclef) [inline middleware/caller.go]:
```
return func
   return func
```

cries in Javascript 😭

File: middleware/caller_test.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package middleware_test
+
+import (
+	"github.com/gol4ng/logger/middleware"
```
Comment (rmasclef) [inline middleware/caller_test.go]: sort imports

File: middleware/context_test.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
+package middleware_test
+
+import (
+	"github.com/gol4ng/logger/middleware"
```
Comment (rmasclef) [inline middleware/context_test.go]: sort imports

File: middleware/context_test.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/gol4ng/logger"
+)
+
+func TestContext_Handle(t *testing.T) {
+	mockHandler := func(entry logger.Entry) error {
```
Comment (rmasclef) [inline middleware/context_test.go]: maybe declare that mock at line 37 for a little more readability ?

File: middleware/filter_test.go [hunk 8]
```diff
@@ -0,0 +1,185 @@
+package middleware_test
+
+import (
+	"fmt"
+	"github.com/gol4ng/logger/middleware"
```
Comment (rmasclef) [inline middleware/filter_test.go]: sort imports

File: ctx.go [hunk 9]
```diff
@@ -0,0 +1,16 @@
+package logger
+
+import "context"
+
+var ContextKey struct{}
```
[the remark this answers is not in the document]
Comment (instabledesign) [reply ctx.go]: https://github.com/golang/go/blob/82ee4e7f78bf34da31e1d050529ab82dc03cf13a/src/context/context.go#L477
 > i set it private

File: ctx.go [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+func InjectInContext(ctx context.Context, l LoggerInterface) context.Context {
+	return context.WithValue(ctx, ContextKey, l)
+}
+
+func FromContext(ctx context.Context, l LoggerInterface) LoggerInterface {
+	if _logger, ok := ctx.Value(ContextKey).(LoggerInterface); ok {
```
[the remark this answers is not in the document]
Comment (instabledesign) [reply ctx.go]: it come from `interface{}` you must cast element to know what you get.

File: middleware/caller.go (hunk 3 above)
Comment (instabledesign) [reply middleware/caller.go]: i renamed it `MiddlewareInterface`

**Pull request merged**